### PR TITLE
Added Snapshot rule resource

### PR DESCRIPTION
**Pull request by mkmayank-dell**
# Description
A few sentences describing the overall goals of the pull request's commits.

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| |

# Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [ ] I have verified that new and existing unit tests pass locally with my changes
- [ ] I have not allowed coverage numbers to degenerate
- [ ] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] Backward compatibility is not broken

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

### CODE REVIEW

File: powerstore/resource_snapshotrule.go [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+				Description:         "The ID of the snapshot rule.",
+				MarkdownDescription: "The ID of the snapshot rule.",
+			},
+			"name": {
+				Type:                types.StringType,
+				Required:            true,
+				Description:         "The name of the snapshot rule.",
+				MarkdownDescription: "The name of the snapshot rule.",
+			},
+			"interval": {
```
Comment (shenda1) [inline powerstore/resource_snapshotrule.go:37]: Is this param mutually exclusive with time_of_day? If yes, please specify in the description.

File: powerstore/resource_snapshotrule.go [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+	}
+
+	if len(plan.DaysOfWeek.Elems) > 0 {
+		snapshotRuleCreate.DaysOfWeek = []gopowerstore.DaysOfWeekEnum{}
+		for _, d := range plan.DaysOfWeek.Elems {
+			snapshotRuleCreate.DaysOfWeek = append(snapshotRuleCreate.DaysOfWeek, gopowerstore.DaysOfWeekEnum(d.String()))
+		}
+	}
+
+	// Add validation
```
Comment (shenda1) [inline powerstore/resource_snapshotrule.go]: Please check if manual validation is required

File: powerstore/resource_snapshotrule.go [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+			"Error getting snapshot rule after creation",
+			"Could not get snapshot rule, unexpected error: "+err.Error(),
+		)
+		return
+	}
+
+	state := models.SnapshotRule{}
+	updateSnapshotRuleState(&state, getRes)
+
+	log.Printf("Added to result: %v", state)
```
Comment (shenda1) [inline powerstore/resource_snapshotrule.go]: log statement needed?

File: powerstore/resource_snapshotrule.go [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+
+	diags = resp.State.Set(ctx, state)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	log.Printf("Done with Create")
+}
+
+// Read resource information
```
Comment (shenda1) [inline powerstore/resource_snapshotrule.go]: Read snapshot rule

File: powerstore/resource_snapshotrule.go (hunk 2 above)
Comment (mkmayank-dell) [reply powerstore/resource_snapshotrule.go]: nope, server give back the error , so removing any type of validation for api

File: powerstore/resource_snapshotrule.go (hunk 3 above)
Comment (mkmayank-dell) [reply powerstore/resource_snapshotrule.go]: nope, just added in case debug , removing

File: powerstore/resource_snapshotrule.go [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+}
+
+// NewResource is a wrapper around provider
+func (r resourceSnapshotRuleType) NewResource(_ context.Context, p tfsdk.Provider) (tfsdk.Resource, diag.Diagnostics) {
+	return resourceSnapshot{
+		p: *(p.(*Pstoreprovider)),
+	}, nil
+}
+
+type resourceSnapshot struct {
```
Comment (shenda1) [inline powerstore/resource_snapshotrule.go]: resourceSnapshotRule will be more relevant

File: powerstore/provider.go [hunk 6]
```diff
@@ -185,6 +187,7 @@ func (p *Pstoreprovider) GetSchema(ctx context.Context) (tfsdk.Schema, diag.Diag
 // this helper can be skipped and the provider type can be directly type
 // asserted (e.g. provider: in.(*provider)), however using this can prevent
 // potential panics.
+//
 //lint:ignore U1000 used by the internal provider, to be checked
```
Comment (shenda1) [inline powerstore/provider.go:191]: if comment is not required, please remove it

File: examples/main.tf [hunk 7]
```diff
@@ -26,6 +26,13 @@ provider "powerstore" {
   insecure = true
 }
 
+
+resource "powerstore_snapshotrule" "test" {
```
Comment (shenda1) [inline examples/main.tf:30]: Please include example with all the user-configurable attributes

File: powerstore/provider.go (hunk 6 above)
Comment (mkmayank-dell) [reply powerstore/provider.go:191]: this new line is result of gofmt

File: powerstore/resource_snapshotrule.go (hunk 5 above)
Comment (mkmayank-dell) [reply powerstore/resource_snapshotrule.go]: Yeah, fixing it

**Review by Sakshi-dell — APPROVED** [review]
create/update snapshot_rule.md file in docs for attributes description.

File: powerstore/resource_snapshotrule.go [hunk 8]
```diff
@@ -0,0 +1,352 @@
...
+				MarkdownDescription: "The interval between snapshots taken by a snapshot rule.",
+			},
+			"time_of_day": {
+				Type:                types.StringType,
+				Optional:            true,
+				Computed:            true,
+				Description:         "The time of the day to take a daily snapshot, with format hh:mm, mutually exclusive with interval parameter.",
+				MarkdownDescription: "The time of the day to take a daily snapshot, with format hh:mm.",
+			},
+			"timezone": {
```
Comment (Sakshi-dell) [inline powerstore/resource_snapshotrule.go:68]: Need to add default attribute for timezone. Default value is "UTC"

Comment (mkmayank-dell) [reply powerstore/resource_snapshotrule.go:68]: this is by default, and server take care of it

Comment (mkmayank-dell) [reply powerstore/resource_snapshotrule.go:68]: also, I cannot send it with every request as timezone can only be sent if time_of_day is provided and not other mutually exclusive interval param

File: powerstore/provider.go (hunk 6 above)
Comment (shenda1) [reply powerstore/provider.go:191]: I am talking about this line "lint:ignore U1000 used by the internal provider, to be checked"

File: powerstore/resource_snapshotrule.go [hunk 9]
```diff
@@ -0,0 +1,352 @@
...
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Get snapshot details from API and then update what is in state from what the API returns
+	id := state.ID.Value
+	response, err := r.p.client.PStoreClient.GetSnapshotRule(context.Background(), id)
+
+	// todo distnguish whether error is for resource presence, in case resource is not present
```
Comment (shenda1) [inline powerstore/resource_snapshotrule.go:357]: comment required?

File: powerstore/resource_snapshotrule.go [hunk 10]
```diff
@@ -0,0 +1,352 @@
...
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Error deleting snapshotRule",
+			"Could not delete snapshotRuleID "+snapshotRuleID+": "+err.Error(),
+		)
+		return
+	}
+
+	// Remove resource from state
+	resp.State.RemoveResource(ctx)
```
Comment (shenda1) [inline powerstore/resource_snapshotrule.go]: Please check if this statement is required

File: powerstore/resource_snapshotrule.go (hunk 8 above)
Comment (mkmayank-dell) [reply powerstore/resource_snapshotrule.go:68]: fixed

File: powerstore/resource_snapshotrule.go (hunk 9 above)
Comment (mkmayank-dell) [reply powerstore/resource_snapshotrule.go:357]: yes, this is to let us know that this is to be done

File: powerstore/resource_snapshotrule.go (hunk 10 above)
Comment (mkmayank-dell) [reply powerstore/resource_snapshotrule.go]: no need, as it will be called automatically ,removing it

**Comment by mkmayank-dell** [comment]
> create/update snapshot_rule.md file in docs for attributes description.

updated

**Review by shenda1 — APPROVED** [review]

**Pull request merged**